imap_processing/ultra/l2/ultra_l2.py

```diff
@@ -0,0 +1,303 @@
+    if output_map_structure.tiling_type is ena_maps.SkyTilingType.HEALPIX:
+        map_nside, map_nested = (
+            output_map_structure.nside,
+            output_map_structure.nested,
+        )
+    else:
+        map_nside, map_nested = (DEFAULT_L2_HEALPIX_NSIDE, DEFAULT_L2_HEALPIX_NESTED)
+
+    # Initialize the HealpixSkyMap object
+    skymap = ena_maps.HealpixSkyMap(
+        nside=map_nside,
+        nested=map_nested,
+        spice_frame=output_map_structure.spice_reference_frame,
+    )
+
+    # Add additional data variables to the map
+    output_map_structure.values_to_push_project.extend(
+        [
+            "observation_time",
+            "pointing_set_exposure_times_solid_angle",
+            "num_pointing_set_pixel_members",
+        ]
+    )
+
+    # Get full list of variables to push to the map: all requested variables plus
+    # any which are required for L2 processing
+    value_keys_to_push_project = list(
+        set(output_map_structure.values_to_push_project + REQUIRED_L1C_VARIABLES)
+    )
+
+    for ultra_l1c_pset in ultra_l1c_psets:
+        pointing_set = ena_maps.UltraPointingSet.from_path_or_dataset(ultra_l1c_pset)
+        logger.info(
+            f"Projecting a PointingSet with {pointing_set.num_points} pixels "
+            f"at epoch:{pointing_set.epoch}\n"
+            f"These values will be projected: {value_keys_to_push_project}"
+        )
+
+        pointing_set.data["num_pointing_set_pixel_members"] = xr.DataArray(
+            np.ones(pointing_set.num_points, dtype=int),
+            dims=(CoordNames.HEALPIX_INDEX.value),
+        )
+        pointing_set.data["observation_time"] = xr.DataArray(
+            np.full(pointing_set.num_points, pointing_set.epoch),
+            dims=(CoordNames.HEALPIX_INDEX.value),
+        )
+        # Add solid_angle * exposure of pointing set as data_var
+        # so this quantity is projected to map pixels for use in weighted averaging
+        pointing_set.data["pointing_set_exposure_times_solid_angle"] = (
+            pointing_set.data["exposure_factor"] * pointing_set.solid_angle
+        )
+
+        # Initial processing for weighted quantities at PSET level
+        # Weight the values by exposure and solid angle
+        pointing_set.data[
+            VARIABLES_TO_WEIGHT_BY_POINTING_SET_EXPOSURE_TIMES_SOLID_ANGLE
+        ] *= pointing_set.data["pointing_set_exposure_times_solid_angle"]
+
+        skymap.project_pset_values_to_map(
+            pointing_set=pointing_set,
+            value_keys=value_keys_to_push_project,
+            index_match_method=ena_maps.IndexMatchMethod.PUSH,
+        )
+
+    # Subsequent processing for weighted quantities at SkyMap level
+    for (
+        quantity_to_weight
+    ) in VARIABLES_TO_WEIGHT_BY_POINTING_SET_EXPOSURE_TIMES_SOLID_ANGLE:
+        skymap.data_1d[quantity_to_weight] = (
+            skymap.data_1d[quantity_to_weight]
+            / skymap.data_1d["pointing_set_exposure_times_solid_angle"]
```

Comment:
Does this one not work like the above?

```python
skymap.data_1d.data[VARIABLES_TO_WEIGHT_BY_POINTING_SET_EXPOSURE_TIMES_SOLID_ANGLE] /= skymap.data_1d.data["pointing_set_exposure_times_solid_angle"]
```

Reply:
Yes! I think this works here. Thanks for catching it.